### DISCUSSION TITLE
Is there any way to paint sharp paths (i.e. without blurry edges)?

### DISCUSSION THREAD

**Discussion by AdeelTariq** [TokisanGames/Terrain3D#562] [Questions] [answered]
I'm trying to paint paths on the terrain but they look like this:
![image](https://github.com/user-attachments/assets/9872cf44-1d6d-4663-b0c5-1617f4d9a45c)

I'd rather they looked sharper like this:
![image](https://github.com/user-attachments/assets/032731bb-57ad-4e4a-86a7-7c95c426954c)

Did I make my terrain too small or something? Or is it just not possible because its vertex based painting? Any other workaround for the kind of look I want to achieve? 
I'm trying to make terrain look similar to A short hike
![image](https://github.com/user-attachments/assets/d37a95e5-f31f-4633-b9a1-c49c4ce3974d)

**Comment by TokisanGames** [comment, marked as answer]
https://terrain3d.readthedocs.io/en/latest/docs/introduction.html#vertex-painting

This terrain is a vertex painter, not a pixel painter. Every terrain pixel is a blend of 4 vertices. You could sharpen things by disabling the blending noise and increasing blend sharpness. Otherwise, you'll have to edit the shader to adjust how it blends between the vertices how you want, such as treating the vertices as points on a Bezier curve or something non-trivial.

**Reply by AdeelTariq** [reply]
Thanks for the reply. I guess I'm out of luck then.
However, I was wondering how autoshader is producing sharper edges based on height? For my specific needs, since I only have like three to four textures, If I can somehow tell the autoshader to paint 3 textures instead of 2 based on height? Somehow tell the autoshader to draw the paths etc.

Anyway thanks for your time. This was just a prototype. So I don't think I'll delve further into this.

**Reply by TokisanGames** [reply]
The shader is editable, so you can see and change how the autoshader works.

**Reply by AdeelTariq** [reply]
The shader in the material section doesn't seem to be handling the blending for painted textures. It only effects the autoshader it seems.
Could you point me to the exact shader or part of the shader that handles the texture painting blending?

![image](https://github.com/user-attachments/assets/2b9888ba-0ab3-41e7-9e1d-93ffd559636d)

**Comment by TokisanGames** [comment]
> The shader in the material section doesn't seem to be handling the blending for painted textures. It only effects the autoshader it seems.

All of the material parameters effects areas painted manually and autoshaded, except for the ones prefaced with "auto", which is only for the autoshader.

> Could you point me to the exact shader or part of the shader that handles the texture painting blending?

Edit the shader in your picture.
Read the documentation https://terrain3d.readthedocs.io/en/stable/docs/shader_design.html
Read the code and comments in this highlighted section https://github.com/TokisanGames/Terrain3D/blob/main/src/shaders/main.glsl#L343-L402

**Reply by AdeelTariq** [reply]
Thanks. I'll give it a try.
